File: scripts/reflecta.rb
if Object.const_defined?(:MRUBY_VERSION)
    $hexdigest = SHA1.method(:sha1_hex)
else
    require 'json'
    require 'digest/sha1'
    $hexdigest = Digest::SHA1.method(:hexdigest)
end

if not $init and $LOAD_PATH
    $init = true
    $LOAD_PATH.each do |path|
        Dir.glob(File.join(path, '*.{rb}')).each do |file|
            if file =~ /(gems|irb|line|bundler|uri|csv|psych|mkmf|resolv|yarp|yaml)/; next; end
            begin; require file; rescue; end
        end
    end
end

class Reflecta
    def self.crash()
        Process.kill('ABRT', Process.pid)
    end

    def self.print(str)
        if $fuzzout == nil
            if ENV["fuzzout"]
                $fuzzout = File.open(ENV["fuzzout"], "w")
            else
                $fuzzout = $stdout
            end
        end

        $fuzzout.puts(str)
        $fuzzout.flush
    end

    def self.filter(b)
        begin
            return eval(b)
        rescue
            return false
        end
    end

    def self.builtins()
        builtins = []
        ObjectSpace.each_object do |o|
            if o.is_a?(Module) and not o.ancestors.include?(Exception) and o.respond_to?(:name) and o.name != nil
                builtins.append(o.name)
            end
        end

        return builtins.select { |b| self.filter(b) }
    end

    def self.enumerate()
        self.print(JSON.dump(self.builtins()))
    end

    def self.record(obj, index)
        self.print(JSON.dump(self.reflect(obj, index)[0]))

    end

    def self.reflect(obj, index)
        return [
            "index" => index,
            "kind" => self.kindof(obj),
            "type" => self.typeof(obj),
            "arity" => self.arityof(obj),
            "fields" => self.fieldsof(obj),
            "methods" => self.methodsof(obj),
        ]
    end


    def self.kindof(obj)
        if [Numeric, String, NilClass, TrueClass, FalseClass].any? { |c| obj.is_a?(c) }
            return "primitive"
        elsif obj.respond_to?(:new) and obj.instance_of?(Class)
            return "class"
        elsif obj.respond_to?(:call)
            return "function"
        else
            return "object"
        end
    end

    def self.typeof(obj)
        case self.kindof(obj)
        when "primitive"
            return obj.class.name
        when "class", "function"
            if obj.respond_to?(:name) and obj.name != nil
                return self.kindof(obj) + "." + obj.name
            else
                return self.kindof(obj)
            end
        when "object"
            if obj.class.respond_to?(:name) and obj.class.name != nil
                return "object." + obj.class.name
            else
                return "object"
            end
        end
    end

    def self.arityof(fn)
        if self.kindof(fn) == "class"
            fn = fn.method(:new)
        end
        if fn.respond_to?(:parameters)
            return fn.parameters.filter {|p| p[0] == :req}.size
        else
            return nil
        end
    end

    def self.methodsof(obj)
        if obj.respond_to?(:methods)
            return obj.methods.map {|m| [m, self.arityof(obj.method(m)).to_s]}
        else
            return []
        end
    end

    def self.fieldsof(obj)
        if obj.respond_to?(:constants)
            return obj.constants
        else
            return []
        end
    end
end

# c = (Object.constants.select { |c| !c.to_s.end_with?("Error") }).to_set
# b = Reflecta.builtins.to_set
# puts b - c

File: scripts/reflecta.php
<?php

if ($init == null) {
    $reflecta = new class() {
        var $fuzzout = null;

        function crash() {
            $pid = getmypid();
            exec("kill -9 $pid");
        }

        function print($s) {
            if ($this->fuzzout == null) {
                if (empty(getenv("fuzzout"))) {
                    $this->fuzzout = STDOUT;
                } else {
                    $this->fuzzout = fopen(getenv("fuzzout"), "w");
                }
            }
            fwrite($this->fuzzout, $s . "\n");
            fflush($this->fuzzout);
        }

        function builtins() {
            $fns = get_defined_functions();
            $cls = get_declared_classes();

            return array_merge($fns["internal"], $cls);
        }

        function enumerate() {
            $this->print(json_encode($this->builtins()));
        }

        function reflect($obj, $index) {
            return [
                "index" => $index,
                "kind" => $this->kindof($obj),
                "type" => $this->typeof($obj),
                "arity" => $this->arityof($obj),
                "fields" => $this->fieldsof($obj),
                "methods" => $this->methodsof($obj),
            ];
        }

        function record($obj, $index) {
            $this->print(json_encode($this->reflect($obj, $index)));
        }

        function kindof($obj) {
            $ty = gettype($obj);
            if ($ty == "string") {
                if (class_exists($obj)) {
                    return "class";
                } else if (function_exists($obj)) {
                    return "function";
                } else {
                    return "primitive";
                }
            } else if ($ty == "object") {
                return "object";
            } else {
                return "primitive";
            }
        }

        function typeof($obj) {
            switch ($this->kindof($obj)) {
                case "primitive":
                    return gettype($obj);
                case "object":
                    return (new ReflectionObject($obj))->getName();
                case "class":
                    return "class." . (new ReflectionClass($obj))->getName();
                case "function":
                    return (new ReflectionFunction($obj))->getName();
            }
        }

        function arityof($fn) {
            switch ($this->kindof($fn)) {
                case "class":
                    return (new ReflectionClass($fn))->getConstructor()?->getNumberOfRequiredParameters();
                case "function":
                    return (new ReflectionFunction($fn))->getNumberOfRequiredParameters();
            }

            return null;
        }

        function fieldsof($obj) {
            $h = function($p) {
                return $p["name"];
            };
            switch ($this->kindof($obj)) {
                case "object":
                    return array_map($h, (new ReflectionObject($obj))->getProperties());
                case "class":
                    return array_map($h, (new ReflectionClass($obj))->getProperties());
                default:
                    return [];
            }
        }


        function methodsof($obj) {
            $h = function ($m) {
                return [$m->getName(), (string) $m->getNumberOfRequiredParameters()];
            };

            switch ($this->kindof($obj)) {
                case "object":
                    return array_map($h, (new ReflectionObject($obj))->getMethods());
                case "class":
                    return array_map($h, (new ReflectionClass($obj))->getMethods());
                default:
                    return [];
            }
        }
    };

    foreach ($reflecta->builtins() as $b) {
        define($b, $b);
    }

    $init = true;
}
